### ISSUE TITLE
[feature request] Offer code mark styles and gutter marks

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/vscode#9209]
_From @darkred on July 10, 2016 23:30_

_(Using VSCode 1.3.0 in win 10 x64 build 10586 with vscode-eslint extension 1.10.18)._

Currently the code marks(for errors and warnings) appear only as squiggly underline (i.e. as **spellchecker errors** in browsers, e.g. Firefox, Chrome):
_(the error is in red, and the warning is in green)_
![2016-07-11_020259](https://cloud.githubusercontent.com/assets/723651/16716650/b635a444-470b-11e6-82fc-90770e8bea05.png)

My suggestion is to offer more code mark styles (ideally  the **outline** style, as it highlights the errors/warnings the best way -see below- ) and **gutter marks** :
_(with the errors to be in red, and the warnings to be in yellow, as ESLint itself uses these colors - not green)_
![2016-07-11_020203](https://cloud.githubusercontent.com/assets/723651/16737733/59dedc70-479b-11e6-8da9-0789ade17143.png)
([here is another screenshot](https://sublimelinter.readthedocs.io/en/latest/_images/marks-outline-dark.png))

It's more easy to distinguish the errors this way.

Thank you

---

_For reference SublimeLinter3 package offers [five different code mark styles](https://sublimelinter.readthedocs.io/en/latest/mark_styles.html) :_ 
* fill, 
* outline   _(the default style)_,
* solid underline, 
* squiggly underline, and 
* stippled underline

_and two types of gutter marks:_ 
* errors, and 
* warnings

_Copied from original issue: Microsoft/vscode-eslint#99_

**Comment by dbaeumer** [comment]
_From @AndrewIngram on July 11, 2016 9:15_

I found myself searching around for a setting to enable this, I guess i've been heavily spoiled by Atom and Sublime. This is pretty important for me, because I find the squiggly lines to be far too difficult to see when scrolling through a file.

**Comment by dbaeumer** [comment]
_From @glebec on July 11, 2016 18:28_

I second the use of yellow for warning. For one thing, red-green is always a poor UI choice as it ignores the 7%+ of XY-chromosomal people who have red-green color blindness. For another, it matches ESLint. Also yellow just feels more like a warning than green, culturally speaking.

Also, the gutter marks can be very helpful, though in the case of VSC it might require some different style as VSC's breakpoint markers are very similar to SublimeLinter error marks.

**Comment by dbaeumer** [comment]
Moving to vscode since such a feature has to be implemented in the editor.

**Comment by dbaeumer** [comment]
@egamma @bpasero I also like the suggestion of using yellow for warning instead of green. Green in dark themes is very hard to read.

**Comment by bpasero** [comment]
I think we only use green because VS does? Not sure...

**Comment by darkred** [comment]
A note about my suggestion on code mark styles:
<br>

I've noticed that Visual Studio (all versions) show errors as squiggly underline as well.
Sp, because VSCode is based on Visual Studio Online's editor (codename "Monaco"), 
I suppose that my suggestion of offering new code mark styles is out of the question (i.e. to deviate from a style that such a flagship product such as Visual studio uses).

But, honestly, showing errors/warnings as squiggly underline is very difficult to distinguish when scrolling through a file (often I have a hard time locating them), 
plus, that this kind of style suggests more to that is a spellcheck error.
e.g.. in the screenshot:
![](https://cloud.githubusercontent.com/assets/723651/16716650/b635a444-470b-11e6-82fc-90770e8bea05.png)
both `while` _(error)_ and `console` _(warning)_ seem to be just **typos**, misspelled words - not the result of static analysis.

The best would be the user to be able to choose the style he prefers.
(IMHO the `outline` style on the other hand, highlights visually the errors/warnings perfectly.)

**Comment by glebec** [comment]
> The best would be the user to be able to choose the style he prefers.

Agreed — I, for example, have a different preference from you in that I like "squiggly" over "outline."

**Comment by egamma** [comment]
@bpasero yes green was picked for consistency with VS. Need to check what VS does in the dark theme @bgashler1 ?

**Comment by stevencl** [comment]
I believe VS uses a more vibrant green:

![image](https://cloud.githubusercontent.com/assets/1704059/17176302/e67e3f80-5403-11e6-9f37-b2dd342f5bd6.png)

**Comment by dbaeumer** [comment]
For me it is not only readability. Green conveys the wrong information as well. And the warning icon is usually a yellow triangle with an exclamation mark. So yellow would be IMO the better color for this even if VS uses a more vibrant green.

**Comment by stevencl** [comment]
I contacted the VS team to learn more about the reason for choosing green. Originally it was yellow (VS2003 I think) but not many people could recall why it was changed. The general consensus though was that it was probably because of the poor contrast that yellow has on a light background.

Yellow definitely makes more sense but we will need to see if we can come up with a colour that looks right and has appropriate contrast in the light theme. @bgashler1 - can you help choose an appropriate colour?

**Comment by glebec** [comment]
@stevencl is there any specific reason we cannot do as @darkred suggests and let warning & error colors be user-defined, or even theme-overridable?

**Comment by bgashler1** [comment]
(This comment is regarding only the color, not the styles...)

A while ago, Microsoft Word set a precedent for using red squiggly lines for errors in spelling and green for warnings of possible grammar mistakes.  This is likely why Visual Studio followed suite, because it was a pattern that people were familiar with in other software and the concepts were similar.

To make the software more inclusive for color blindness, recent versions of Word now use blue instead of green for warnings (see below).

![capture](https://cloud.githubusercontent.com/assets/11839736/17422127/de4fe6c2-5a63-11e6-9287-95411e25e325.PNG)

I propose we use blue here for warnings.  I find blue to work well in both light and dark settings.

**Comment by bgashler1** [comment]
@glebec As far as whether we can do user-defined styles or theme-defined styles for errors and warning, I will put this on our UX sync agenda to get our team's feedback.  Thanks for the great suggestion!

**Comment by coreh** [comment]
If the setting is added, it would be really nice for it to support the "stippled" / dotted underline, since that look is the native look for the spell checker on Mac OS X

![](https://cloudup.com/c5CXxRhjWkp+)

In fact it should probably be the default look on OS X, if no option is specified.

**Comment by gold** [comment]
I just switched from Atom to VS Code about a week ago. I have a short list of things that need to fixed. This issue is one of them. Arguing over which color is right or wrong will never solve the problem for everyone. Choose reasonable defaults as best you can (green seems weird). But allow the user to override the default styles. Additionally, optionally allowing corresponding error/warning glyphs in the gutter is a great idea. Also while I'm in rant mode, expressing the theme in XML is so old fashioned. I know it's a plist thing, but honestly, do you have to slavishly copy Apple even when it's wrong?

**Comment by DavidBabel** [comment]
> For reference SublimeLinter3 package offers five different code mark styles :
> 
> fill,
> outline (the default style),
> solid underline,
> squiggly underline, and
> stippled underline

For me this is one of the main thing i miss in vscode. 

It can be a good idea to consider to add another mark style like a "red color" for the text in error.
That's the case in IntelliJ, as you can see in this video tutorial for typescript : https://youtu.be/n6RoVyZEsv4?t=13m20s
![capture d ecran 2016-10-24 a 15 47 48](https://cloud.githubusercontent.com/assets/5599375/19648102/43d6bfe8-9a01-11e6-8d91-24353a39641c.png)


Each time an error occurs, the text become red.

Does anyone can confirm me it's not possible yet ?

**Comment by javiayala** [comment]
+1
I would love to see a yellow/red mark in the gutter like in atom. The green squiggly underline is just hard to see and easy to miss. I need a way to customize this.

**Comment by DavidBabel** [comment]
Some outline actually exists in vscode atm. Like when you target a bracket and it highlight the corresponding one : 

<img width="106" alt="capture d ecran 2016-11-04 a 10 34 46" src="https://cloud.githubusercontent.com/assets/5599375/20001023/8edfb036-a27a-11e6-99f3-4dad1e2b0b6e.png">

Don't know if it can be used elsewhere

**Comment by donaldpipowitch** [comment]
Could this issue be tackled in multiple steps to see faster progress?
E.g. switch warning color from green to yellow (or blue?) in the first step. Than allow different error/warning highlights like dotted lines, gutter marks, outlines, etc.

**Comment by rohan07** [comment]
Any update on this?

**Comment by proProbe** [comment]
Would love to have this! I agree that green is not the best color to use when it comes to warnings. So any progress on this?

**Comment by Ikarium** [comment]
Please, squiggly are very ugly, especially if you run linters.

**Comment by rednithin** [comment]
Green is a show stopper for warnings.

**Comment by tannerlinsley** [comment]
This would be such a huge improvement.  Almost every other editor I've used has this and it's the most helpful thing ever :) @waderyan

**Comment by rohan07** [comment]
@bgashler1 - Any update?

**Comment by DHedgecock** [comment]
It is definitely possible to configure the colors for linter errors/warnings as of the last release 🎉🎉🎉. You can use these variables in your preferences:

Errors and warnings:

`editorError.foreground`: Foreground color of error squigglies in the editor.
`editorError.border`: Border color of error squigglies in the editor.
`editorWarning.foreground`: Foreground color of warning squigglies in the editor.
`editorWarning.border`: Border color of warning squigglies in the editor.

like:

```javascript
"workbench.colorCustomizations": {
    "editorWarning.foreground": "#336699",
    "editorWarning.border": "#336699"
  }
```

If you have a favorite theme, you can also put in a pull request to have them add it in the theme. I've done this in the Panda Extended theme:
https://github.com/DHedgecock/vscode-panda-extended/blob/master/dist/Panda%20Extended.json

![screen shot 2017-06-21 at 3 34 56 pm](https://user-images.githubusercontent.com/8461733/27409627-488f07fe-5697-11e7-94c9-2f14441505df.png)

**Comment by glebec** [comment]
At long last! So our current status is as follows:

* [x] error line colors
* [ ] error line styles
* [ ] error gutter marks

IMHO the first was the most important, the others are niceties which I think justify keeping this issue open but which aren't as big quality of life improvements.

**Comment by DHedgecock** [comment]
Actually it looks like it's also possible to have gutter icons now:
<img width="931" alt="screen shot 2017-06-21 at 3 50 28 pm" src="https://user-images.githubusercontent.com/8461733/27410087-e759126a-5699-11e7-8143-e8ef3262f052.png">

^ that screenshot is from pulling down the codemetrics extension and making this change to it:
<img width="1136" alt="screen shot 2017-06-21 at 3 54 59 pm" src="https://user-images.githubusercontent.com/8461733/27410107-02f4589a-569a-11e7-94a3-564274b70637.png">

Docs for `gutterIconPath` and `gutterIconSize` available here: https://code.visualstudio.com/docs/extensionAPI/vscode-api#_a-namedecorationrenderoptionsaspan-classcodeitem-id279decorationrenderoptionsspan

Worth doing some investigation into.

**Comment by glebec** [comment]
So what we'd want to do is add a setting to VSC to display gutter marks for lines with errors or warnings, so we don't rely on extensions (e.g. ESLint) to implement this behavior individually – yes? The plumbing is there, in theory creating a new setting for it would be not too bad. There would be some built-in images for the marks. As a further goal, themes could optionally provide additional ones, though that's lower priority.

**Comment by Ikarium** [comment]
Additional marker style also, dotted lines for ex., box around the strings, ability to change colour/background colour.
For the moment we have error and warning, maybe add some other: Info, spellchecks?

**Comment by aeschli** [comment]
Warning and info squigglies and borders now use separate color ids:
`editorWarning.foreground`, `editorWarning.border`,
`editorInfo.foreground`, `editorInfo.border`
That means you can now customize these colors in the user settings or in a theme.

I haven't changed the default colors yet: warning and info are still green. Any color recommendations from the UX team. @stevencl ?

**Comment by lukewhitehouse** [comment]
Hey @aeschli ! Is there an ETA on when these options will be available in the live app? I'm very keen to start using them like many others :)

**Comment by aeschli** [comment]
`editorWarning.foreground`, `editorWarning.border`,
`editorInfo.foreground`, `editorInfo.border` are all in the latest stable release (1.17)

**Comment by taka4sato** [comment]
Hi @aeschli , thank you very much for the support, I really love this feature.
Can I confirm is there any way to highlight (apply different color) just only for eslint errors/warnings?

I think setting `editorError.foreground:` will also override the normal syntax errore which cause build fails.

**Comment by aeschli** [comment]
@taka4sato Sorry, the squiggly coloring is based on the problem's severity (Error, Warning, Info) and there's currently no way of having different colors per problem type.

**Comment by mrpbennett** [comment]
are we any closer to having the ability to change squiggly lines for dots in the gutter? Or does the above enable us to do that, or is just the ability to change colour?

**Comment by avondohren** [comment]
Any update to this topic? I love the ability to update the squiggly color, but having a glyph or marking in the gutter would really help visually see that an action is needed on that line.

Thanks in advance! Loving VS Code so far!

**Comment by tannerlinsley** [comment]
This is still a huge item for me :)

**Comment by serhiipalash** [comment]
VS Code / SublimeText

<img width="1677" alt="screen shot 2018-03-18 at 12 03 10 pm" src="https://user-images.githubusercontent.com/14208343/37564759-f66d0862-2aa4-11e8-8390-65367efc94db.png">

Compare how `no-inline-styles` rule warning looks in Sublime and VS Code and check how minimaps differ. 

I think this is the biggest lack in VS Code. :(

**Comment by linhtto** [comment]
Referencing https://github.com/Microsoft/vscode/issues/3915 which is also open.

**Comment by ianizaguirre** [comment]
I found a way to change the error red squiggly line into an outline style (kind of like Sublime Text), so errors are much easier to spot! I like outlined errors MUCH better over small squiggly lines, because I find the squiggly lines to be difficult to see when scrolling through a file. 

So this was my end result in VS Code:

![vscode-squiggly-error-to-outline-error](https://user-images.githubusercontent.com/29879270/48297096-a3c60800-e46e-11e8-8593-7b2b5b4a0a43.png)

I wrote down the steps I followed to get this to work [HERE](https://www.izaguir.re/how-to-make-vscode-error-highlighting-look-more-like-sublime-error-highlighting/)

**Comment by tannerlinsley** [comment]
This is genius. I was just complaining about this again today.

**Comment by ianizaguirre** [comment]
@tannerlinsley Thanks! I literally have spent weeks searching trying to figure this out because I found it so annoying that I had to squint to find errors lol Glad it could help you.

**Comment by PinkaminaDianePie** [comment]
created extension for gutters, it's easier than to wait for another 2 years: https://marketplace.visualstudio.com/items?itemName=IgorSbitnev.error-gutters&fbclid=IwAR3jyNd4wbZUz8m36wMRlxUwcpGFCqcbadhEQpuZsyKzbqIw_UL_HGRnddU

**Comment by gold** [comment]
thank you!

btw, I've mostly switched to vscode from atom.

Atom overall ux is better then vscode. But vscode performance is
significantly better. and the vscode built-in terminal is amazing. (I'm a
Linux user btw)

**Comment by trusktr** [comment]
This is such a crucial feature. Coming from other editors, it is just straight difficult to spot errors while scrolling. Looking at the error ticks in the scroll bar and trying to match the little line with the marks is not so convenient either.

This is the first major feature I miss, coming from Atom. Besides that, I really like VS Code.

**Comment by onehorsetown** [comment]
@trusktr 

Try this:  https://marketplace.visualstudio.com/items?itemName=IgorSbitnev.error-gutters

**Comment by agentlogic** [comment]
In addition to the above extension, I found this pretty useful: [Error Lens](https://marketplace.visualstudio.com/items?itemName=usernamehw.errorlens)
